feat(alert,notificationdrawer): add support to truncate title

boaz0 opened the pull request:
**What**: Closes #4357 

**Additional issues**:
 - Do not merge till `@patternfly/patternfly` is updated to `4.10.33`

//cc @mcoker

patternfly-build commented:
PF4 preview: https://patternfly-react-pr-4399.surge.sh

dlabrecq commented on packages/react-core/src/components/Alert/Alert.tsx:
```diff
@@ -48,9 +51,18 @@ export const Alert: React.FunctionComponent<AlertProps & OUIAProps> = ({
+    }
+    titleRef.current.style.setProperty(maxLines.name, truncateTitle.toString())
+  }, [titleRef, truncateTitle])
+
```
Curious why you chose to use a `titleRef` Vs applying the modifier directly to `className`?

boaz0 replied:
@dlabrecq the use of `titleRef` is to inject a new value to the `*-title--max-lines` variable which indicates the number of lines to show in the title.

I could not find any way to set it as `className`. If you know a way to do so, let me know. :smiley_cat:

mcoker commented:
Does this add support for setting `--pf-c-notification-drawer__group-toggle-title--max-lines` to limit the number of lines shown in `.pf-c-notification-drawer__group-toggle-title`? Thats an element we don't have a truncate modifier for (it truncates by default), but we still support the var to change the number of lines of truncation.

mcoker commented:
This lgtm! Also to follow up on my original question, it looks like if you pass `0` as a value to the max-lines var, the text will not truncate. So I think you could apply the `truncate=number` prop to `.pf-c-notification-drawer__group-toggle-title`, too, and it would set `--pf-c-notification-drawer__group-toggle-title--max-lines` (without setting `.pf-m-truncate`, like the prop on the other components), and `0` could be used to allow the title not to trucnate.

boaz0 commented:
Hi @mcoker 
- I added support to `--pf-c-notification-drawer__group-toggle-title--max-lines` but it seems not to work - mind taking a look what I am missing? Maybe I set it on the wrong element? You can see it in the group example of the docs.
- I updated the default value to `0` as you suggested.

mcoker commented:
@boaz0 looks like the title element is missing the class `pf-c-notification-drawer__group-toggle-title`

boaz0 commented:
resolved conflict.

dlabrecq commented:
@boaz0 Looks like you're still working on @mcoker 's suggestion? Let me know and I'll review, again.

karelhala commented on packages/react-core/src/components/Alert/Alert.tsx:
```diff
@@ -48,10 +51,19 @@ export const Alert: React.FunctionComponent<AlertProps> = ({
   ouiaSafe = true,
   ...props
 }: AlertProps) => {
+  const titleRef = React.useRef(null);
```
Unfortunately we can't use hooks in PF components due to older versions of React.

boaz0 replied:
No problem, I don't have to use Hooks for that. :+1:

tlabaj replied:
As of the breaking change release, we updated our React dependency to `react@^16.8.0` so we can now use hooks.

karelhala commented on packages/react-core/src/components/Alert/Alert.tsx:
```diff
@@ -48,10 +51,19 @@ export const Alert: React.FunctionComponent<AlertProps> = ({
+    if (!titleRef.current || !truncateTitle) {
+      return;
+    }
+    titleRef.current.style.setProperty(maxLines.name, truncateTitle.toString());
```
Also might be better to limit just the `title` from props in `getHeadingContent` if it's a string. Otherwise consumer has to do the truncation.

boaz0 replied:
I am not sure what you meant. But I cannot do that in `getHeadingContent` as it is wrapping `title` in `React.Fragment`:

```
    <React.Fragment>
      <span className={css(accessibleStyles.screenReader)}>{variantLabel}</span>
      {title}
    </React.Fragment>
```

I guess I misunderstood you. Can you give an example?

karelhala replied:
Here you could trim the `title` if I am reading the code correctly. But as I said, that's my preference to trim the string instead of css rule. But both are doable!

mcoker commented on packages/react-core/src/components/NotificationDrawer/NotificationDrawerGroup.tsx:
```diff
@@ -32,28 +35,39 @@ export const NotificationDrawerGroup: React.FunctionComponent<NotificationDrawer
+          aria-expanded={isExpanded}
+          onClick={e => onExpand(e, !isExpanded)}
+        >
+          <div ref={titleRef}>{title}</div>
```
I think this needs to be 

```suggestion
          <div ref={titleRef} className={css(styles.notificationDrawerGroupToggleTitle)}>{title}</div>
```

tlabaj commented:
@boaz0 looks like you still need to make the update the @mcoker suggested.  Can you please make that change and resolve the conflicts.  Thanks!

patternfly-build commented:
Your changes have been released in: 
 - @patternfly/react-catalog-view-extension@4.8.15
 - @patternfly/react-core@4.45.0
 - @patternfly/react-docs@5.8.18
 - @patternfly/react-inline-edit-extension@4.5.71
 - demo-app-ts@4.33.2
 - @patternfly/react-table@4.16.4
 - @patternfly/react-topology@4.4.72
 - @patternfly/react-virtualized-extension@4.5.60

Thanks for your contribution! :tada:
